### DISCUSSION TITLE
Standardised warnings & agnostic warning handling à la `np.errstate`?

### DISCUSSION THREAD

**Discussion by lucascolley** [data-apis/array-api#849] [General]
https://data-apis.org/array-api/2023.12/design_topics/exceptions.html#exceptions hints that some of what arises here is out of scope for the standard, but opening this discussion to give concrete examples of some problems I have hit.

---

Working in the array API standard ecosystem I have bumped into two complications related to warnings.

Firstly, there is a long-standing Dask issue opened by @shoyer regarding NumPy warnings escaping from warning context managers when Dask arrays delay computation - https://github.com/dask/dask/issues/3245.

Arguably this is a Dask bug and is irrelevant to the standard, but nevertheless it is causing errors when trying to make SciPy work with all standard(-ish) libraries. Ideally, for SciPy, there would be an array-agnostic way to capture that we want to 'ignore warnings about divide by 0 in this section of code' regardless of the implementation of the underlying library `xp`.

Related to the above problem - https://github.com/data-apis/array-api-extra/issues/14 proposes finding a public home for SciPy's private helper `_lazywhere` in array-api-extra. The helper is written to be array-agnostic, apart from a use of `with np.errstate(invalid='ignore'):`. Functions in array-api-extra are strictly array-agnostic, depending only on the standard.

This isn't the end of the world for SciPy, as we could instead just wrap `xpx.lazywhere` in a light `np.errstate` context manager. But this is obviously less ideal than being able to maintain the granularity of the context manager. And complications could arise if other implementations to NumPy start throwing their own warnings too.

**Comment by rgommers** [comment]
> Arguably this is a Dask bug

It's something Dask could fix, but also I've always considered these particular "empty slice" warnings a design problem in NumPy. The warnings are quite annoying in a lot of contexts - if the correct (or agreed-upon) return value for `np.mean([])` is `nan`, then no warning by default is a lot nicer. I can't really think of other functions other than the `mean` & co reductions that warn like this.

**Comment by seberg** [comment]
FWIW, I don't mind changing it in NumPy.  And we probably just should!

The question may be for any function where I think we return `NaN`, but a normal array would raise if it was an empty slice.  For those, the warnings actually seem fine to me (and a `default=` kwarg might be the way to get around it).

**Reply by rgommers** [reply]
Are there any such functions? `mean`, `std`, `var`, median` at least all do return `nan` for the non-nan functions.

**Reply by seberg** [reply]
Even if not, there could be.  But yes: `np.nanmin` and `np.nanmax`.

**Reply by seberg** [reply]
(I suppose those actually already raise in NumPy, so it doesn't matter)

**Comment by lucascolley** [comment]
Actually, I think this isn't a problem for `lazywhere` since `where` does not support Python scalar `x2` yet. And when it does (gh-807), `result_type` should do too, so the workaround that introduced the NumPy warning should no longer be needed.
